### PR TITLE
[mlir][mesh] Add lowering of process multi-index op

### PR DESCRIPTION
**Pull request by sogartar**
* Rename mesh.process_index -> mesh.process_multi_index.
* Add mesh.process_linear_index op.
* Add lowering of mesh.process_multi_index into an expression using mesh.process_linear_index, mesh.cluster_shape and affine.delinearize_index.

This is useful to lower mesh ops and prepare them for further lowering where the runtime may have only the linear index of a device/process. For example in MPI we have a rank (linear index) in a communicator.

### CODE REVIEW

**Comment by sogartar** [comment]
@yaochengji, will you take a look at this PR?

File: mlir/test/Dialect/Mesh/ops.mlir [hunk 1]
```diff
@@ -156,26 +156,26 @@ func.func @cluster_shape_empty_axes() -> (index, index, index) {
   return %0#0, %0#1, %0#2 : index, index, index
 }
 
-// CHECK-LABEL: func @process_index
-func.func @process_index() -> (index, index) {
-  // CHECK: %[[RES:.*]]:2 = mesh.process_index on @mesh0 axes = [0, 1] : index, index
-  %0:2 = mesh.process_index on @mesh0 axes = [0, 1] : index, index
+// CHECK-LABEL: func @process_multi_index
```
Comment (yaochengji) [inline mlir/test/Dialect/Mesh/ops.mlir:159]: ops/invalid check of mesh.process_linear_index seems missing?

Comment (sogartar) [reply mlir/test/Dialect/Mesh/ops.mlir:159]: I added tests.

**Review by yaochengji — APPROVED** [review]

**Comment by yaochengji** [comment]
LGTM, thanks

**Comment by sogartar** [comment]
Just rebased and resolved conflicts.

**Pull request merged**